storage: add delayed acks for snapshots

bdarnell opened the pull request:
We were previously calling raft.RawNode.ReportSnapshot as soon as we
passed the snapshot message to the transport, which defeated raft's
internal flow control and could cause pending snapshot messages to pile
up if they could not be completely transmitted in less than the
heartbeat interval.

This commit adds a channel to the RaftTransport object that can be used
to report the status of the snapshot as soon as the last of the data has
been sent.

Fixes #5709

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5721)

<!-- Reviewable:end -->

bdarnell committed:
storage: add delayed acks for snapshots

We were previously calling raft.RawNode.ReportSnapshot as soon as we
passed the snapshot message to the transport, which defeated raft's
internal flow control and could cause pending snapshot messages to pile
up if they could not be completely transmitted in less than the
heartbeat interval.

This commit adds a channel to the RaftTransport object that can be used
to report the status of the snapshot as soon as the last of the data has
been sent.

Fixes #5709

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[storage/raft_transport.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-KE7d9gWamqNCHOLwLvG:-KE7d9gWamqNCHOLwLvH:-625944987) ([raw file](https://github.com/cockroachdb/cockroach/blob/47553d4be3597c8cea930570b6ef4b9c98a6a1cf/storage/raft_transport.go#L256)):*
doesn't seem like this is needed since processRaft won't bail until this task completes.

---

*[storage/raft_transport_test.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-KE7do3AcwWiXOmTmez7:-KE7do3BKXyUI50M61Bh:-598003715) ([raw file](https://github.com/cockroachdb/cockroach/blob/47553d4be3597c8cea930570b6ef4b9c98a6a1cf/storage/raft_transport_test.go#L334)):*
s/Inof/Info/, s/call/calling/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-:-KE7e5OZNISrtCeZpC5_:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/raft_transport_test.go, line 377 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-KE7hqn4f1R9QJTiF0cS:-KE7hqn4f1R9QJTiF0cT:-843224778) ([raw file](https://github.com/cockroachdb/cockroach/blob/47553d4be3597c8cea930570b6ef4b9c98a6a1cf/storage/raft_transport_test.go#L377)):*
instead of maintaining numSnaps and numMessages, couldn't you range over msgTypes?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/raft_transport.go, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-KE7d9gWamqNCHOLwLvG:-KE7oxSWk7tFcCSOYWaQ:-452322293) ([raw file](https://github.com/cockroachdb/cockroach/blob/47553d4be3597c8cea930570b6ef4b9c98a6a1cf/storage/raft_transport.go#L256)):*
Good point. I added this because of a test failure, but it was just one of the `RaftTransport` tests that wasn't using a real store. I've removed this select and `TestSendAndReceive` is now aware of `SnapshotStatusChan`.  The new test is now basically redundant so I've removed it.

---

*[storage/raft_transport_test.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-KE7do3AcwWiXOmTmez7:-KE7pIHt2QO4-ZOVsc2I:-587748681) ([raw file](https://github.com/cockroachdb/cockroach/blob/47553d4be3597c8cea930570b6ef4b9c98a6a1cf/storage/raft_transport_test.go#L334)):*
This line is gone, but I fixed it in the test I copied it from.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721#-:-KE7q9iE0-WcSEUPb8k-:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5721)*
<!-- Sent from Reviewable.io -->
